### PR TITLE
Type member and static field edits take fast path

### LINKED ISSUE #1906: [meta] address all `disable-fast-path` tests
Any test with `disable-fast-path` is going to probably cause anomalous behavior in LSP. We've got a number of remaining tests that have the fast path disabled; this is a meta-issue tracking the remaining issues that require us to disable fast path tests:

### Errors arising from not repeating `typeMember` passes in the incremental resolver:
- test/testdata/core/fuzz_bad_subtyping.rb
- test/testdata/resolver/type_members.rb
- test/testdata/resolver/type_member_missing.rb
- test/testdata/resolver/fuzz_type_member_forget.rb
- test/testdata/infer/bad_child_class.rb
- test/testdata/resolver/redefinition_of_subclass_type_member.rb
- test/testdata/resolver/enumerable_strict.rb
- test/testdata/resolver/type_member_missing_then_use.rb
- test/testdata/infer/generics/self_params.rb

### `is not a class or module` not raised in fast path
- test/testdata/namer/nested_static_field.rb
- test/testdata/infer/bad_suggest.rb

### `unable to resolve right-hand side` not raised in fast path
- test/testdata/autogen/generator.rb
- test/testdata/resolver/fuzz_alias_to_type_alias.rb
- test/testdata/resolver/bad_alias.rb

### `redefining constant` not raised in fast path (#1907)
- test/testdata/namer/constant_redefinition/constant_then_class.rb
- test/testdata/namer/constant_redefinition/constant_then_module.rb
- test/testdata/namer/class_and_alias.rb

### wrong expected type in fast path
- test/testdata/resolver/overloads_test.rb

### duplicated `only modules can be included` errors in fast path
- test/testdata/resolver/linearization/includes_class.rb

### infinite types or excessively recursive are weird
- test/testdata/resolver/fuzz_infinite_type.rb
- test/testdata/resolver/infinite.rb
Note: this last one might not be robustly fixable, as it involves different cutoff points that we stop reporting a ridiculous number of errors, but we otherwise report largely the same errors between them.

### type member/alias errors don't get re-reported in the fast-path
- test/testdata/dsl/opus_enum_alias.rb
- test/testdata/infer/generics/bounds.rb
- test/testdata/infer/generics/bounds_super.rb
- test/testdata/resolver/type_member_cycle.rb
- test/testdata/resolver/type_member_fixed_order.rb
- test/testdata/resolver/type_member_bad_parent.rb

### PR DESCRIPTION
**Pull request by jez**
### Background

A handful of changes landed in prep for this that may be helpful to know about:

- #6440
- #6435
- #6421
- #6418
- #6411
- #6472
- #6471
- #6518
- #6538
- #6540
- #6534
- #6521

~The last two are probably the most important.~ #6418 defined all classes before all type members and static fields, so that in this PR there can be a clear separation between defining classes, deleting static fields and type members, and then defining new static fields and type members. (and then #6411 was arguably a pre-work change to landing #6418 itself).

### Summary

With those background changes out of the way (most of which were no-op's), this change is a pretty straightforward continuation of changes like #5808 or #6278.

I was at first thinking that we would want to do static fields and type members separately for the sake of going slowly and deliberately, but the mangle-rename class of errors that arise from the interaction between these two symbol kinds are easier to solve by doing them all at once.

Also note that this doesn't handle **all** static fields, as static field class alises (`A = B`) are still not handled. Class aliases suffer from the same class of problem mentioned in #5941, but I fear that because class aliases are more widespread, that they could cause incorrect incremental behavior more frequently than the `alias_method` cases, so I've punted on them for the time being. It's worth noting that we already track `classalias` and `onlyclassalias` changed definition metrics, so we can track the impact to fixing that particular class of edits going forward.

As another benefit, this finally rids us of all known problems with the fast path:

Fixes #1906

(In a future change, we may want to delete the `# disable-fast-path: true` code and require all tests to pass in the fast path. Though it's possible we want to keep it just in case.)

### Commit summary

Most of the changes on this branch can be reviewed in standalone commits. See the Commits tab.

### Test plan

I've written some tests. Some more on the way, but still ready for review

- [x] `@@foo` is also a static field
- [x] Something with `T.attached_class` (`<AttachedClass>` is a type_member)
- [x] When you do `type_member` on fast path, be sure to handle `type_template`, because we always create a class alias from the instance class to the singleton class, and that might screw things up.
- [x] Test for changing type of fixed type_member on fast path
  - [x] Test for changing fixed to not fixed on fast path
- [x] Add a test with `class << self; type_member ...; end`
- [x] Test case: rename static-field → type-member → static-field → type-member
- [x] Take the all_constant_redefinitions namer test and make it into multiple fast path tests
- [x] Really double check type_member fast path changes, because they're not
  going to be exercised commonly common in pay-server (most type_member usages
  have tons of extraFiles implicated)
- [x] Do we need/want a remove_static_field_from_generic_class.rb test?
- [x] Test that reordering two type members of generic class checks downstream
  - [x] Should there be some sort of "shape hash" for type members that looks its index up in its owner?
  - [x] Seems like maybe `reorder_type_members` passes without needing any extra work like this because in any case where the new order is visible, it would also have the effect of changing a method's signature, and so change some method's hash which affects which extraFiles to typecheck.
  (If in the future we find an example where that's not the case, we could of course add a type member shape has or something.)
- [x] ~~Tests for a type_member that has a scope (should not be allowed; change scope name, change part after the scope)~~
- [x] Test for scope with static field.
  - If it comes down to it we _might_ be able to skip these symbols in the hash function, but it would be tricky. Probably easier to support it outright, hopefully
  - The problem with scopes is that they store `ClassOrModuleRef` IDs, which might not be stable (instead of indices into the `definedClasses` vector that namer keeps track of)
  - This is maybe not hard now that we define FoundClass's for all static field scopes?
- [x] Test case: variations of
  ```ruby
  X = 1
  X::Y = 1
  ```
  - [x] You added one, but you can keep adding more.
- [x] It looks like changes to static field class aliases are taking the fast path? Is that a problem?
- [x] ~~Change to name of type member needs to retypecheck all files mentioning
  enclosing class, too~~
  - Realistically, in Stripe's codebase any change to a type member is going to
    take the slow path until we build the incremental slow path, so let's just
    revisit then.
- [x] I still have a few TODOs in the implementation which I plan to take a look at.
- [x] You're still technically deleting and redefining static field class
  aliases; it just so happens that they don't change semantically when doing so
  (so things won't resolve differently), but they will get different IDs. Can
  you write a test case that would force you to have to skip over
  `isClassAlias` symbols in the `deleteSymbolViaFullNameHash` method?
  - I didn't end up doing this. Hope that doesn't lead to problems.

#### Later

- [ ] We should probably make class variables show up in static fields, rather than fields?

### CODE REVIEW

**Comment by jez** [comment]
Current dependencies on/for this PR:
* master
  * **PR #6418** <a href="https://app.graphite.dev/github/pr/sorbet/sorbet/6418" target="_blank">Graphite</a> 
    * **PR #6421** <a href="https://app.graphite.dev/github/pr/sorbet/sorbet/6421" target="_blank">Graphite</a> 
      * **PR #6422** <a href="https://app.graphite.dev/github/pr/sorbet/sorbet/6422" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sorbet/sorbet/6422?utm_source=stack-comment).

**Review by froydnj — DISMISSED** [review]
This doesn't feel surprising, excellent work on the pre-work PRs.

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_MWsnd59kSNM0H7
→ https://go/builds/bui_MWsnBCZvGjJ2QJ

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_MgFDJYwfYVV6Ct
→ https://go/builds/bui_MgFDgfmT68Wi1L

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_MiuiR8IThEuk7i
→ https://go/builds/bui_Miui0iXCznBNMP

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_Mk2lXxjtloHdUd
→ https://go/builds/bui_Mk2lRUoUZQwGI3

File: test/testdata/lsp/fast_path/add_static_field_to_generic_class.1.rbupdate [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# typed: true
+# assert-fast-path: add_static_field_to_generic_class.rb
```
[the remark this answers is not in the document]
Comment (froydnj) [reply test/testdata/lsp/fast_path/add_static_field_to_generic_class.1.rbupdate:2]: > Do we need/want a `remove_static_field_from_generic_class.rb` test?

Ping on this comment.  I think there's an add/remove test for type members further down; do we also want a add/remove static field + generic class test?  (These cross product tests can get kind of annoying...)

**Review by froydnj — APPROVED** [review]
Small comments, but I don't think anything here is a hard blocker.

File: core/FoundDefinitions.h [hunk 2]
```diff
@@ -210,7 +210,7 @@ CheckSize(FoundField, 20, 4);
 class FoundDefinitions final {
     // Contains references to items in _staticFields and _typeMembers.
     // Used so there is a consistent definition & redefinition ordering.
-    std::vector<FoundDefinitionRef> _nonDeletableDefinitions;
+    std::vector<FoundDefinitionRef> _deletableDefinitions;
```
[the remark this answers is not in the document]
Comment (jez) [reply core/FoundDefinitions.h:213]: Yes we definitely do. I'll punt this to a followup

**Pull request merged**